rust/lance/src/dataset/sql.rs

```diff
@@ -0,0 +1,171 @@
+#[derive(Clone, Debug)]
+pub struct SqlBuilder {
+    /// The dataset to run the SQL query
+    pub(crate) dataset: Option<Dataset>,
+
+    /// The SQL query to run
+    pub(crate) sql: String,
+
+    /// the name of the table to register in the datafusion context
+    pub(crate) table_name: Option<String>,
+
+    /// If true, the query result will include the internal row id
+    pub(crate) row_id: bool,
+
+    /// If true, the query result will include the internal row address
+    pub(crate) row_addr: bool,
+}
+
+impl SqlBuilder {
+    /// The table name to register in the datafusion context.
+    /// This is used to specify a "table name" for the dataset.
+    /// So that you can run SQL queries against it.
+    /// If not set, the default table name is "dataset".
+    pub fn table_name(mut self, table_name: &str) -> Self {
+        self.table_name = Some(table_name.to_string());
+        self
+    }
+
+    /// Specify if the query result should include the internal row id.
+    /// If true, the query result will include an additional column named "_rowid".
+    pub fn with_row_id(mut self, row_id: bool) -> Self {
+        self.row_id = row_id;
+        self
+    }
+
+    /// Specify if the query result should include the internal row address.
+    /// If true, the query result will include an additional column named "_rowaddr".
+    pub fn with_row_addr(mut self, row_addr: bool) -> Self {
+        self.row_addr = row_addr;
+        self
+    }
+
+    pub async fn execute(self) -> lance_core::Result<QueryResult> {
+        let ctx = SessionContext::new();
+        ctx.register_table(
+            self.table_name.unwrap(),
+            Arc::new(LanceTableProvider::new(
+                Arc::new(self.dataset.unwrap()),
+                self.row_id,
+                self.row_addr,
+            )),
+        )?;
+        let df = ctx.sql(&self.sql).await?;
+        Ok(QueryResult::new(df))
+    }
+}
+
+impl Default for SqlBuilder {
+    fn default() -> Self {
+        Self {
+            dataset: None,
+            sql: "".to_string(),
+            table_name: Some("dataset".to_string()),
+            row_id: false,
+            row_addr: false,
+        }
+    }
+}
+
+pub struct QueryResult {
+    dataframe: DataFrame,
+}
+
+impl QueryResult {
+    pub fn new(dataframe: DataFrame) -> Self {
+        Self { dataframe }
+    }
+
+    pub async fn into_stream(self) -> SendableRecordBatchStream {
+        self.dataframe.execute_stream().await.unwrap()
+    }
+
+    pub async fn collect(self) -> lance_core::Result<Vec<RecordBatch>> {
+        use futures::TryStreamExt;
+        Ok(self
+            .dataframe
+            .execute_stream()
+            .await
+            .unwrap()
+            .try_collect::<Vec<_>>()
+            .await?)
+    }
+
+    pub fn dataframe(&self) -> &DataFrame {
+        &self.dataframe
+    }
+
+    pub async fn explain(&self, verbose: bool, analyze: bool) -> DataFrame {
```

Comment:
I think it's probably better to also consume self, make this `into_explain_plan(self, ...)`, because for example if it is analyze, it actually runs the plan.